server/util/cosmos.js

```diff
@@ -13,225 +19,122 @@ import { PUBSUB_TOPIC_MISC } from '../constant';
+      const delegatorAddress = firstAccount.address;
+      const tendermint34Client = await Tendermint34Client.connect(cosmosRPCEndpoint);
+      const queryClient = QueryClient.withExtensions(
+        tendermint34Client,
+        setupAuthExtension,
+        setupBankExtension,
+      );
+      const signingClient = await SigningStargateClient.connectWithSigner(
+        cosmosRPCEndpoint, wallet,
+      );
+      return { delegatorAddress, queryClient, signingClient };
+    })();
   }
-  // eslint-disable-next-line no-console
-  console.error(`${likecoin.denom} is not supported denom`);
-  return -1;
+  return cosmosPromise;
 }
 
-export function getCosmosDelegatorAddress() {
-  return cosmosSigner.cosmosAddress;
+export async function getCosmosDelegatorAddress() {
+  const { delegatorAddress } = await getCosmos();
+  return delegatorAddress;
 }
 
 export async function getCosmosAccountLIKE(address) {
-  const { data } = await api.get(`/auth/accounts/${address}`);
-  if (!data.result.value || !data.result.value.coins || !data.result.value.coins.length) return 0;
-  const likecoin = data.result.value.coins.find((c) => c.denom === COSMOS_DENOM);
-  return likecoin ? amountToLIKE(likecoin) : 0;
-}
-
-export async function getAccountInfo(address) {
-  const res = await api.get(`/auth/accounts/${address}`);
-  if (res.data.result) {
-    return res.data.result.value;
-  }
-  return res.data.value;
+  const { queryClient } = await getCosmos();
+  const { amount } = await queryClient.bank.balance(address, COSMOS_DENOM);
+  return amount;
 }
 
-function signTransaction({
-  signer,
-  accNum,
-  stdTx: inputStdTx,
-  sequence,
-}) {
-  const stdTx = { ...inputStdTx };
-  const signMessage = jsonStringify({
-    fee: stdTx.fee,
-    msgs: stdTx.msg,
-    chain_id: COSMOS_CHAIN_ID,
-    account_number: accNum,
-    sequence: sequence.toString(),
-    memo: stdTx.memo,
-  });
-  const { signature, publicKey } = signer.sign(Buffer.from(signMessage, 'utf-8'));
-  stdTx.signatures = [{
-    signature: signature.toString('base64'),
-    account_number: accNum,
-    sequence: sequence.toString(),
-    pub_key: {
-      type: 'tendermint/PubKeySecp256k1',
-      value: publicKey.toString('base64'),
-    },
-  }];
-  return stdTx;
+async function getAccountInfo(address) {
+  const { queryClient } = await getCosmos();
+  const { value } = await queryClient.auth.account(address);
+  const accountInfo = BaseAccount.decode(value);
+  return accountInfo;
 }
 
 async function sendTransaction(signedTx) {
-  try {
-    const res = await api.post('/txs', {
-      tx: signedTx,
-      mode: 'sync',
-    });
-    if (res.data.code) {
-      throw new Error(res.data.raw_log);
-    }
-    return res.data.txhash;
-  } catch (err) {
-    // eslint-disable-next-line no-console
-    console.error(err);
-    if (err && err.response && err.response.data) {
-      if (err.response.data.error
-        && err.response.data.error.includes('Tx already exists')) {
```

Comment:
Can we handle the tx exists case in the new code?
Like https://github.com/likecoin/likecoin-jedi/pull/51/files#diff-889cd002a3740ec9ed5df105b2d50359050ec2772898562502a3ddf4d3ea1180R289

Reply:
Curious about in what circumstance will we have a existed tx?

Reply:
And will we prefer to use the same way to manually calculate the `txHash` and return?

Reply:
When the transaction was submitted, but somehow the txHash was not stored into database.
In this case the transaction will be retried, but the same transaction already exists in the node's txpool, an tx exists error would be thrown

Reply:
Only when the returned error is `transaction already exists` we would want to calculate the hash offline

Reply:
Updated